### DISCUSSION TITLE
mutation with variables yield null logical issues on the graphql-java side

### DISCUSSION THREAD

**Discussion by guykatz** [graphql-java/graphql-java#2388] [Q&A]
Hi all;
we are facing something that I hope can be explained or fixed somehow.
we are using gql-java 14 on the server and apollo client 2 on the client.
our web applciation uses queries/mutations with variables (pretty basic stuff)
to super simplify what we are facing lets llook at the following example:
schema definition:
in types:
`
type Address{
   street: String
   city: String
}
type Person{
   name:String!
   address: Address
}
`
in mutations...
`
updatePerson(
   name:String!
   address:Address
):Person
`
so we have a person and he has an address. address is nullable both in the type and in the mutation.
our web app code query string for the mutation looks something like:
`
mutation updatePerson($name: String!, $addressConfig: Addres) {...
`
when we execute this mutation without the address (variables contain only name), on the graphql java data fetcher we get a key-value pair "address:null" in the DataFetchingEnvironmentImpl.agruments map.
this is a big problem because there's a difference between **not** getting this key:value pair at all in the env map and **getting** it with null value. the latter is not correct IMHO because nullifying an address is an actual thing I might want to do as oposed to 'not touching' it at all which is my intention. 
so not sending any address in the mutations logically means I do not want to do anything with the exiting address and sending it with null logically means I want to nullify what ever is in the address.
now I am no graphql expert but I think the graphql java implemenation is when this happens.
can anyone comment on this? this really is a big issue for me and I hope I am doing something wrong or that there;'s aconfig option I missed somewhere that will solve this to me.
thanks!

**Comment by guykatz** [comment]
BTW, looks like in 16+ this is solved. its weird that 14 was like this and people worked with it this is why I am still a little baffled and maybe I am missing something

**Reply by bbakerman** [reply]
Early versions of the graphql spec did not specify clearly how to handle missing versus null arguments.

It was later cleared up  BUT the graphql-java code still had bugs in it from the old undefined behavior

This has been fixed in the latest versions and we are more confident in our missing versus explicitly null handling

**Reply by guykatz** [reply]
@bbakerman thanks a lot for your reply and explanation.
IMHO if 14 is a LTS version then this is probably one of the highest importance ones to consider in future 14.x bug fix versions...
I would suggest that logically it is a super important to distinguish between the 2 cases.
thanks again!
